### ISSUE TITLE
Server returns J3 extension versions for requests from J4

### ISSUE THREAD

**Issue by Bakual** [joomla/install-from-web-server#64]
While trying out the Weblinks and also my own extension (SermonSpeaker) it noticed that the Webinstaller offers the J3 version of the packages instead of the J4 ones.
Once installed, you get an update to the J4 version in the extension updater.

I think the server is missing code to properly take care of Joola 4.0 requests.

**Comment by Llewellynvdm** [comment]
This is the one place the version set is managed: 

https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/base.php#L308

To see the JSON that the scrapper works with try this:

**Comment by Llewellynvdm** [comment]
> This is the one place the version set is managed: 

https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/base.php#L308

> To see the JSON that the scrapper works with try this:

view-source:https://extensions.joomla.org/index.php?option=com_jed&view=extension&controller=filter&filter[approved]=1&filter[published]=1&filter[core_catid]=129&format=json

> Here is the return set

http://jsoneditoronline.org/#left=cloud.de35690427884e9ea011d178cc9e43ab

> The call is made here per category:

https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/category.php#L151

> The call per extension:

https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/extension.php#L69

Still not sure where to issues is, but at a quick glance this is what I found....

**Comment by Llewellynvdm** [comment]
Download link if its is the XML update server, gets given to the CMS install extension... so me thinking out loud would think the issue is there.

So just to confirm you are using an XML update server as your direct link on the JED, right?

**Comment by Bakual** [comment]
I honestly don't know what Weblinks is using, but for SermonSpeaker I use the XML (https://raw.githubusercontent.com/Bakual/SermonSpeaker/main/updates.xml), yes.
![image](https://user-images.githubusercontent.com/1018684/134979560-2949609e-6c5d-4f55-a707-d5c0c0e83ab7.png)

**Comment by Llewellynvdm** [comment]
Okay so once you select the extension in the webinstall, it pass that XML to the install extension.... can you check this?

**Comment by Bakual** [comment]
No, it doesn't pass the XML, it gets the URL to the ZIP and passes that one. The XML is not returned to the plugin.
![image](https://user-images.githubusercontent.com/1018684/134980074-fdde1c05-e24a-415f-a649-942dac70f200.png)

**Comment by Llewellynvdm** [comment]
Can you check this with [weblinks](https://extensions.joomla.org/extension/official-extensions/weblinks/)?

**Comment by Llewellynvdm** [comment]
Hmmm I see it does the same... so it could be the client... still looking.

**Comment by Llewellynvdm** [comment]
Could be right here... https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/extension.php#L175

Not sure but that does look like the moment the download url is build.... wonder if the [cache](https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/extension.php#L148) API makes differentiation between Joomla 3 and 4?

**Comment by Llewellynvdm** [comment]
And since this website APP is running on a Joomla 3 website the [Update class](https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/extension.php#L179) only gets what will install on it.

**Comment by Llewellynvdm** [comment]
Yes I see the filter.dev_level... wonder if it is being passed correctly...

**Comment by Llewellynvdm** [comment]
![image](https://user-images.githubusercontent.com/5607939/134983193-7c88084f-d891-463a-b957-b2410b07227b.png)

**Comment by Llewellynvdm** [comment]
SO nope there is the issue:
![image](https://user-images.githubusercontent.com/5607939/134983373-83744cf3-0857-4e24-b225-9046a5a0473d.png)

`Mw==` is 3

So I conclude at this point that this is a client issue.

**Comment by Llewellynvdm** [comment]
This is where the next search on the client side has started:
https://github.com/joomla-extensions/install-from-web-client/blob/master/plugins/installer/webinstaller/webinstaller.php#L134

**Comment by Llewellynvdm** [comment]
Okay so we could run a test on the server side, by [changing these few lines](https://github.com/joomla/install-from-web-server/blob/master/components/com_apps/models/extension.php#L175):
```php

if (preg_match('/\.xml\s*$/', $item->downloadurl))
{
	$release = $this->getState('filter.release');

	$update = new Update;
	$update->set('jversion.release', $release);
	$update->loadFromXML($item->downloadurl);
	$packageUrlNode = $update->get('downloadurl', false);

	if (isset($packageUrlNode->_data))
	{
		$item->downloadurl = $packageUrlNode->_data;
	}
}
```
I am not sure if that will work, but @richard67 thinks the release property contains the version not the dev level, and I would agree with him.

**Comment by Llewellynvdm** [comment]
Well it could be because the end-user may not know that the XML link will work...

But okay, lets look at that as the solution as I can see the wisdom in that. We don't touch the URL in the app, we pass the xml instead. I will try to test that later today.

**Comment by Bakual** [comment]
Tested it with Weblinks in a J3 and a J4 site and it worked fine. For my SermonSpeaker it also installed the right version.
